### PR TITLE
Configuration format v1

### PR DESCRIPTION
**Pull request by neilalexander**
This PR contains the v1 configuration format for Dendrite ahead of the beta. It is a fairly large departure from the `dendrite-config.yaml` of today, instead breaking up the config as much as possible by component. 

Still to do:
- [ ] Check the format makes sense as far as possible
- [ ] Update sample configs in repo (including Docker)
- [ ] Do `listen` and `bind` want to be changed to something more descriptive?
- [ ] matrix-org/sytest#921

Generated sample config:
```
version: 1
global:
  server_name: localhost
  private_key: matrix.pem
  key_validity_period: 168h0m0s
  trusted_third_party_id_servers: []
  kafka:
    addresses: []
    use_naffka: true
    naffka_database:
      connection_string: file:naffka.db
      max_open_conns: 100
      max_idle_conns: 2
      conn_max_lifetime: -1
    topics:
      output_room_event: OutputRoomEventTopic
      output_client_data: OutputClientDataTopic
      output_typing_event: OutputTypingEventTopic
      output_send_to_device_event: OutputSendToDeviceEventTopic
      output_key_change_event: OutputKeyChangeEventTopic
  metrics:
    enabled: false
    basic_auth:
      username: metrics
      password: metrics
app_service_api:
  listen: localhost:7777
  bind: localhost:7777
  database:
    connection_string: file:appservice.db
    max_open_conns: 100
    max_idle_conns: 2
    conn_max_lifetime: -1
  config_files: []
client_api:
  listen: localhost:7771
  bind: localhost:7771
  registration_shared_secret: ""
  recaptcha_public_key: ""
  recaptcha_private_key: ""
  enable_registration_captcha: false
  captcha_bypass_secret: ""
  recaptcha_siteverify_api: ""
  registration_disabled: false
  turn:
    turn_user_lifetime: ""
    turn_uris: []
    turn_shared_secret: ""
    turn_username: ""
    turn_password: ""
current_state_server:
  listen: localhost:7782
  bind: localhost:7782
  database:
    connection_string: file:currentstate.db
    max_open_conns: 100
    max_idle_conns: 2
    conn_max_lifetime: -1
edu_server:
  listen: localhost:7778
  bind: localhost:7778
federation_api:
  listen: localhost:7772
  bind: localhost:7772
  federation_certificates: []
federation_sender:
  listen: localhost:7775
  bind: localhost:7775
  database:
    connection_string: file:federationsender.db
    max_open_conns: 100
    max_idle_conns: 2
    conn_max_lifetime: -1
  send_max_retries: 16
  disable_tls_validation: false
  proxy_outbound:
    enabled: false
    protocol: http
    host: localhost
    port: 8080
key_server:
  listen: localhost:7779
  bind: localhost:7779
  database:
    connection_string: file:keyserver.db
    max_open_conns: 100
    max_idle_conns: 2
    conn_max_lifetime: -1
media_api:
  listen: localhost:7774
  bind: localhost:7774
  database:
    connection_string: file:mediaapi.db
    max_open_conns: 100
    max_idle_conns: 2
    conn_max_lifetime: -1
  base_path: ./media_store
  max_file_size_bytes: 10485760
  dynamic_thumbnails: false
  max_thumbnail_generators: 10
  thumbnail_sizes: []
room_server:
  listen: localhost:7770
  bind: localhost:7770
  database:
    connection_string: file:roomserver.db
    max_open_conns: 100
    max_idle_conns: 2
    conn_max_lifetime: -1
server_key_api:
  listen: localhost:7780
  bind: localhost:7780
  database:
    connection_string: file:serverkeyapi.db
    max_open_conns: 100
    max_idle_conns: 2
    conn_max_lifetime: -1
  key_perspectives: []
sync_api:
  listen: localhost:7773
  bind: localhost:7773
  database:
    connection_string: file:syncapi.db
    max_open_conns: 100
    max_idle_conns: 2
    conn_max_lifetime: -1
user_api:
  listen: localhost:7781
  bind: localhost:7781
  account_database:
    connection_string: file:userapi_accounts.db
    max_open_conns: 100
    max_idle_conns: 2
    conn_max_lifetime: -1
  device_database:
    connection_string: file:userapi_devices.db
    max_open_conns: 100
    max_idle_conns: 2
    conn_max_lifetime: -1
tracing:
  enabled: false
  jaeger:
    serviceName: ""
    disabled: false
    rpc_metrics: false
    tags: []
    sampler: null
    reporter: null
    headers: null
    baggage_restrictions: null
    throttler: null
logging:
- type: file
  level: info
  params:
    path: /var/log/dendrite
```

### CODE REVIEW

**Comment by Kegsay** [comment]
A few breaking changes I want to see:
 - The big rename of components, with the rule that "external" components use `-api` and "internal" components use `-server`, so specifically:
    * userapi -> userserver
    * serverkeyapi -> serverkeyserver
    * federationsender -> federationsenderserver
    * appservice -> appserviceserver
 - Should we expose topics _at all_? For all intents and purposes, they are random strings we could absolutely hardcode, with maybe a `topic_prefix` option to resolve any clashes? Otherwise, every new topic is a breaking change as we will rely on them to startup.
 - Can we move:
```
  database: file:appservice.db
  database_options:
    database_max_open_conns: 0
    database_max_idle_conns: 0
    database_conn_max_lifetime: -1
```
to look like:
```
  database:
    name: file:appservice.db
    database_max_open_conns: 0
    database_max_idle_conns: 0
    database_conn_max_lifetime: -1
```
so we can share the same generic "databasey" struct around?
- As you mention, rename listen|bind to be something more useful.

File: internal/config/config_clientapi.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	// TURN options
+	TURN TURN `yaml:"turn"`
+}
+
+func (c *ClientAPI) Defaults() {
+	c.Listen = "localhost:7771"
+	c.Bind = "localhost:7771"
+	c.RegistrationSharedSecret = ""
+	c.RecaptchaPublicKey = ""
+	c.RecaptchaPrivateKey = ""
```
Comment (Kegsay) [inline internal/config/config_clientapi.go:43]: These seem redundant?

Comment (Kegsay) [reply internal/config/config_clientapi.go:43]: I think maybe if `UnmarshalYAML` did the default setting as well then that would make things a bit simpler? We wouldn't forget calling `Defaults()` then!

Comment (neilalexander) [reply internal/config/config_clientapi.go:43]: They probably are!

File: build/gobind/monolith.go [hunk 2]
```diff
@@ -156,7 +158,11 @@ func (m *DendriteMonolith) Start() {
 		RoomserverAPI:       rsAPI,
 		UserAPI:             userAPI,
 		StateAPI:            stateAPI,
+<<<<<<< HEAD
 		KeyAPI:              keyserver.NewInternalAPI(&base.Cfg.KeyServer, federation, userAPI, base.KafkaProducer),
+=======
+		KeyAPI:              keyAPI,
+>>>>>>> master
```
Comment (Kegsay) [inline build/gobind/monolith.go]: Probably need to remove this :D

**Pull request merged**